File: src-webservices/code/org/meandre/webservices/servlets/WSRepositoryServlet.py
#
# Implements the basic about services
#

__name__ = 'WSRepositoryServlet'

requestMap = {
    'GET': { 
        'dump': 'repository_dump',    
        'regenerate': 'repository_regenerate',    
        'list_components': 'repository_list_components',    
        'list_flows': 'repository_list_flows',    
        'tags': 'repository_tags',    
        'tags_components': 'repository_tags_components',    
        'tags_flows': 'repository_tags_flows',    
        'components_by_tag': 'repository_components_by_tag',    
        'flows_by_tag': 'repository_flows_by_tag',    
        'describe_component': 'repository_describe_component',    
        'describe_all_components': 'repository_describe_all_components',    
        'describe_flow': 'repository_describe_flow',    
        'describe_all_flows': 'repository_describe_all_flows',    
        'search_components': 'repository_search_components',    
        'search_flows': 'repository_search_flows',    
        'remove': 'repository_remove'
    },
    'POST': { 
        'add': 'repository_add',    
        'add_flow_description': 'repository_add_flow_descriptions' 
    }
}

#
# Required imports
#

from com.hp.hpl.jena.rdf.model import Resource

from org.meandre.core.repository import RepositoryImpl

from org.meandre.webservices.servlets import WSRepositoryServlet

#
# Services implementation
#

def repository_dump ( request, response, format ):
    '''Returns the user repository stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        content = meandre_store.getRepositoryStore(getMeandreUser(request)).getModel()
        statusOK(response)
        sendRDFModel(response,content,format)
    else:
        errorForbidden(response)
        

def repository_regenerate ( request, response, format ):
    '''Regenerates the user repository stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        meandre_store.regenerateRepository(getMeandreUser(request))
        content = {'message':'Repository successfully regenerated'}
        statusOK(response)
        sendTJXContent(response,[content],format)
    else:
        errorForbidden(response)
    
    
def repository_list_components ( request, response, format ):
    '''List the components aggregated in the user repository stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        content, order, limit, ordered = [], 'name', -1, 0
        if 'order' in params :
            ordered = 1
            if params['order'][0]=='date' :
                order = 'date'
        if 'limit' in params :
            try:
                ordered = 1
                limit = int(params['limit'][0])
            except :
                log.warning('Wrong limit format '+str(params['limit'][0])+'. Defaulting to unlimited')
        qr, resources = meandre_store.getRepositoryStore(getMeandreUser(request)), None
        if ordered :
            resources = qr.getAvailableExecutableComponentsOrderedBy(order,limit)            
        else:
            resources = qr.getAvailableExecutableComponents()
        for resource in resources:
            content.append({'meandre_uri':resource.toString()})
        statusOK(response)
        sendTJXContent(response,content,format)
    else:
        errorForbidden(response)
    
     
def repository_list_flows ( request, response, format ):
    '''List the flows aggregated in the user repository stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        content, order, limit, ordered = [], 'name', -1, 0
        if 'order' in params :
            ordered = 1
            if params['order'][0]=='date' :
                order = 'date'
        if 'limit' in params :
            try:
                ordered = 1
                limit = int(params['limit'][0])
            except :
                log.warning('Wrong limit format '+str(params['limit'][0])+'. Defaulting to unlimited')
        qr, resources = meandre_store.getRepositoryStore(getMeandreUser(request)), None
        if ordered :
            resources = qr.getAvailableFlowsOrderedBy(order,limit)            
        else:
            resources = qr.getAvailableFlows()
        for resource in resources:
            content.append({'meandre_uri':resource.toString()})
        statusOK(response)
        sendTJXContent(response,content,format)
    else:
        errorForbidden(response)
    
    
def repository_tags ( request, response, format ):
    '''Returns the tags stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        qr = meandre_store.getRepositoryStore(getMeandreUser(request))
        content = []
        for tag in qr.getTags():
            content.append({'meandre_tag':tag})
        statusOK(response)
        sendTJXContent(response,content,format)
    else:
        errorForbidden(response)
   
   
def repository_tags_components ( request, response, format ):
    '''Returns the component tags stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        qr = meandre_store.getRepositoryStore(getMeandreUser(request))
        content = []
        for tag in qr.getComponentTags():
            content.append({'meandre_tag':tag})
        statusOK(response)
        sendTJXContent(response,content,format)
    else:
        errorForbidden(response)
  
  
def repository_tags_flows ( request, response, format ):
    '''Returns the component tags stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        qr = meandre_store.getRepositoryStore(getMeandreUser(request))
        content = []
        for tag in qr.getFlowTags():
            content.append({'meandre_tag':tag})
        statusOK(response)
        sendTJXContent(response,content,format)
    else:
        errorForbidden(response)
  
def repository_components_by_tag ( request, response, format ):
    '''List the components aggregated in the user repository that match
       the provided tags stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'tag' in params :
            content = []
            qr = meandre_store.getRepositoryStore(getMeandreUser(request))
            for tag in params['tag'] :
                components = qr.getComponentsByTag(tag) 
                if components is not None :
                    for component in components :
                         content.append({
                                'meandre_uri': component.getExecutableComponent().toString()
                            })
            statusOK(response)
            sendTJXContent(response,content,format)
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)
        
  
def repository_flows_by_tag ( request, response, format ):
    '''List the flows aggregated in the user repository that match
       the provided tags stored in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'tag' in params :
            content = []
            qr = meandre_store.getRepositoryStore(getMeandreUser(request))
            for tag in params['tag'] :
                flows = qr.getFlowsByTag(tag)
                if flows is not None : 
                    for component in flows :
                        content.append({
                            'meandre_uri': component.getFlowComponent().toString()
                            })
            statusOK(response)
            sendTJXContent(response,content,format)
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)
        
    
def repository_describe_all_components ( request, response, format ):
    '''Returns all the component descriptions  aggregated for the requesting user in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        qr = meandre_store.getRepositoryStore(getMeandreUser(request))
        content = getEmptyModel()
        for component_desc in qr.getAvailableExecutableComponentDescriptions():
            content.add(component_desc.getModel())
        statusOK(response)
        sendRDFModel(response,content,format)
    else:
        errorForbidden(response)

   
def repository_describe_all_flows ( request, response, format ):
    '''Returns all the flow descriptions aggregated for the requesting user in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        qr = meandre_store.getRepositoryStore(getMeandreUser(request))
        content = getEmptyModel()
        for component_desc in qr.getAvailableFlowDescriptions():
            content.add(component_desc.getModel())
        statusOK(response)
        sendRDFModel(response,content,format)
    else:
        errorForbidden(response)
   
  
def repository_describe_component ( request, response, format ):
    '''Returns all the components aggregated for the requesting user in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'uri' in params :
            qr = meandre_store.getRepositoryStore(getMeandreUser(request))
            content = getEmptyModel()
            for component_uri in params['uri']:
                component_desc = qr.getExecutableComponentDescription(content.createResource(component_uri))
                if component_desc is not None:
                    content.add(component_desc.getModel())
            statusOK(response)
            sendRDFModel(response,content,format)
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)

  
def repository_describe_flow ( request, response, format ):
    '''Returns all the flows aggregated for the requesting user in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'uri' in params :
            qr = meandre_store.getRepositoryStore(getMeandreUser(request))
            content = getEmptyModel()
            for flow_uri in params['uri']:
                flow_desc = qr.getFlowDescription(content.createResource(flow_uri))
                if flow_desc is not None:
                    content.add(flow_desc.getModel())
            statusOK(response)
            sendRDFModel(response,content,format)
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)

def repository_remove ( request, response, format ):
    '''Remove the associated uri element from the aggregated repository
       for the requesting user in the current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'uri' in params :
            qr = meandre_store.getRepositoryStore(getMeandreUser(request))
            user_model = qr.getModel()
            content = []
            for uri in params['uri']:
                resource_uri = user_model.createResource(uri)
                component_desc = qr.getExecutableComponentDescription(resource_uri)
                flow_desc = qr.getFlowDescription(resource_uri)
                model_to_remove = None
                if component_desc is not None :
                    model_to_remove = component_desc.getModel()
                    content.append({'meandre_uri':uri})
                elif flow_desc is not None :
                    model_to_remove = flow_desc.getModel()
                    content.append({'meandre_uri':uri})
                else :
                    model_to_remove = getEmptyModel()
                user_model.begin()
                user_model.remove(model_to_remove)
                user_model.commit()
            qr.refreshCache()
            statusOK(response)
            sendTJXContent(response,content,format)
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)  
    
    
def repository_search_components ( request, response, format ):
    '''Search the components aggregated in the user repository stored in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'q' in params:
            content, order, limit, ordered, q = [], 'name', -1, 0, params['q'][0]
            if 'order' in params :
                ordered = 1
                if params['order'][0]=='date' :
                    order = 'date'
            if 'limit' in params :
                try:
                    ordered = 1
                    limit = int(params['limit'][0])
                except :
                    log.warning('Wrong limit format '+str(params['limit'][0])+'. Defaulting to unlimited')
            qr, resources = meandre_store.getRepositoryStore(getMeandreUser(request)), None
            resources = qr.getAvailableExecutableComponents(q)
            if ordered :
                model = getEmptyModel()
                for resource in resources:
                    model.add(qr.getExecutableComponentDescription(resource).getModel())
                qrNew = RepositoryImpl(model)
                resources = qrNew.getAvailableExecutableComponentsOrderedBy(order,limit)            
            
            for resource in resources:
                content.append({'meandre_uri':resource.toString()})
            statusOK(response)
            sendTJXContent(response,content,format) 
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)

    
def repository_search_flows ( request, response, format ):
    '''Search the components aggregated in the user repository stored in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'q' in params:
            content, order, limit, ordered, q = [], 'name', -1, 0, params['q'][0]
            if 'order' in params :
                ordered = 1
                if params['order'][0]=='date' :
                    order = 'date'
            if 'limit' in params :
                try:
                    ordered = 1
                    limit = int(params['limit'][0])
                except :
                    log.warning('Wrong limit format '+str(params['limit'][0])+'. Defaulting to unlimited')
            qr, resources = meandre_store.getRepositoryStore(getMeandreUser(request)), None
            resources = qr.getAvailableFlows(q)
            if ordered :
                model = getEmptyModel()
                for resource in resources:
                    model.add(qr.getFlowDescription(resource).getModel())
                qrNew = RepositoryImpl(model)
                resources = qrNew.getAvailableFlowsOrderedBy(order,limit)            
            
            for resource in resources:
                content.append({'meandre_uri':resource.toString()})
            statusOK(response)
            sendTJXContent(response,content,format) 
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)
   
        
def repository_add_flow_descriptions ( request, response, format ):
    '''Add a flow descriptor the user repository stored in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        params = extractRequestParamaters(request)
        if 'repository' in params:
            repositories = params['repository']
            overwrites = params['overwrite']
            diff = len(repositories)-len(overwrites)
            if diff>0 :
                overwrite.append(['false' for i in range(diff)])
            user = getMeandreUser(request)
            content = []
            for repository,overwrite in zip(repositories,overwrites) :
                uris = meandre_store.addFlowsToRepository(user, repository, overwrite=='true')
                for uri in uris :
                    content.append({'meandre_uri':uri})
            statusOK(response)
            sendTJXContent(response,content,format) 
        else:
            errorExpectationFail(response)
    else:
        errorForbidden(response)
  
def repository_add ( request, response, format ):
    '''Add components and flows to  the user repository stored in the 
       current Meandre server.'''
    if checkUserRole (request,Role.REPOSITORY) :
        uris = WSRepositoryServlet.addToRepository(request,meandre_store,meandre_config,format)
        content = []
        for uri in uris :
            content.append({'meandre_uri':uri})
        statusOK(response)
        sendTJXContent(response,content,format) 
    else:
        errorForbidden(response) 
    
    

File: src-webservices/code/org/meandre/webservices/servlets/WSRepositoryServlet.java
/**
 *
 */
package org.meandre.webservices.servlets;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.HashSet;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import javax.servlet.http.HttpServletRequest;

import org.apache.commons.fileupload.FileItem;
import org.apache.commons.fileupload.FileUploadException;
import org.apache.commons.fileupload.disk.DiskFileItemFactory;
import org.apache.commons.fileupload.servlet.ServletFileUpload;
import org.meandre.configuration.CoreConfiguration;
import org.meandre.core.repository.ExecutableComponentDescription;
import org.meandre.core.repository.FlowDescription;
import org.meandre.core.repository.QueryableRepository;
import org.meandre.core.repository.RepositoryImpl;
import org.meandre.core.store.Store;
import org.meandre.core.utils.ModelIO;
import org.meandre.plugins.PluginFactory;
import org.meandre.plugins.tools.JarToolServlet;
import org.meandre.plugins.tools.ResourceConflictException;
import org.meandre.support.crypto.Crypto;
import org.meandre.webservices.MeandreServer;
import org.meandre.webservices.logger.WSLoggerFactory;

import com.hp.hpl.jena.datatypes.xsd.XSDDatatype;
import com.hp.hpl.jena.rdf.model.Literal;
import com.hp.hpl.jena.rdf.model.Model;
import com.hp.hpl.jena.rdf.model.ModelFactory;
import com.hp.hpl.jena.rdf.model.Resource;

/** Provides the servlet that serves Repository requests.
 *
 * @author Xavier Llor&agrave;
 *
 */
public class WSRepositoryServlet extends MeandreBaseServlet {

	/** A default serial ID */
	private static final long serialVersionUID = 1L;

	/** The web service logger */
	private static final Logger log = WSLoggerFactory.getWSLogger();

	/** The resource manager that interrogates the JarTool plugin for the existence of JARs based on MD5 checksums */
	private static JarToolServlet resManager;

	/** Creates the servlet to provide Repository information.
	 *
	 * @param server The Meandre server
	 * @param store The Meandre store to use
	 * @param cnf The Meandre configuration to use
	 */
	public WSRepositoryServlet(MeandreServer server, Store store, CoreConfiguration cnf) {
		super(server, store, cnf);
	}

	/** Adds components and flows to the user repository.
	 *
	 * @param request The request object
	 * @param store The Meandre store
	 * @param cnf The Meandre core configuration object
	 * @param sExtension The extension format
	 * @return The set of added URIs
	 * @throws FileUploadException An exception was encountered while uploading the model
	 */
	@SuppressWarnings("unchecked")
	public static final Set<String> addToRepository(HttpServletRequest request, Store store, CoreConfiguration cnf, String sExtension)
	throws IOException, FileUploadException {

		boolean bEmbed = false;
		boolean bOverwrite = false;
		Model modelAcc = ModelFactory.createDefaultModel();
		ServletFileUpload upload = new ServletFileUpload(new DiskFileItemFactory());
		List lstItems = upload.parseRequest(request);
		Iterator<FileItem> itr = lstItems.iterator();
		HashSet<ExecutableComponentDescription> setComponentsToAdd = new HashSet<ExecutableComponentDescription>();
		Hashtable<String,byte[]> htContextBytes = new Hashtable<String,byte[]>();

		// The user repository
		QueryableRepository qr = store.getRepositoryStore(request.getRemoteUser());

		while(itr.hasNext()) {
			FileItem item = itr.next();
		    // Get the name of the field
			String fieldName = item.getFieldName();
			byte[] data = item.get();

			// check if the current item is a form field or an uploaded file
            if(fieldName.equals("repository")) {

				String sRDFContent = new String(data);
				if (sRDFContent.length()==0 ) continue;

				Model modelTmp = ModelFactory.createDefaultModel();

				modelTmp.setNsPrefix("", "http://www.meandre.org/ontology/");
				modelTmp.setNsPrefix("xsd", "http://www.w3.org/2001/XMLSchema#");
				modelTmp.setNsPrefix("rdfs","http://www.w3.org/2000/01/rdf-schema#");
				modelTmp.setNsPrefix("dc","http://purl.org/dc/elements/1.1/");

				//
				// Read the location and check its consistency
				//
				ModelIO.attemptReadModel(modelTmp,sRDFContent);

				//
				// Accumulate the models
				//
				QueryableRepository qrTmp = new RepositoryImpl(modelTmp);
				QueryableRepository qrAcc = new RepositoryImpl(modelAcc);

				for ( ExecutableComponentDescription ecd:qrTmp.getAvailableExecutableComponentDescriptions() )
					if ( !qrAcc.getAvailableExecutableComponents().contains(ecd.getExecutableComponent()) )
						modelAcc.add(ecd.getModel());
					else
						log.warning("Discarding component "+ecd.getExecutableComponent()+". It already exists in the repository");

				for ( FlowDescription fd:qrTmp.getAvailableFlowDescriptions() )
					if ( !qrAcc.getAvailableFlows().contains(fd.getFlowComponent()) )
						modelAcc.add(fd.getModel());
					else
						log.warning("Discarding flow "+fd.getFlowComponent()+". It already exists in the repository");

			}
			// Check if we need to upload jar files
			else if(fieldName.equals("context")) {
				String sFile = item.getName();
				if ( sFile.length()==0) continue;
				htContextBytes.put(sFile, data);
			}
			else if ( fieldName.equals("embed") ) {
				String sValue = item.getString();
				if ( sValue.length()==0 ) continue;
				bEmbed = Boolean.parseBoolean(sValue);
			}
			else if ( fieldName.equals("overwrite") ) {
				String sValue = item.getString();
				if ( sValue.length()==0 ) continue;
				bOverwrite = Boolean.parseBoolean(sValue);
			}
		}

		// TODO: Check that this uploaded contexts are really proper contexts :D
		//
		// Check the uploaded description
		//
		QueryableRepository qrNew = new RepositoryImpl(modelAcc);

		//
		// The resulting set
		//
		Set<String> setAddedURIs = new HashSet<String>();

		synchronized ( qr ) {
			//
			// Add to the user repository
			//
			// Adding components
			for ( Resource resComp:qrNew.getAvailableExecutableComponents() ) {
				if ( qr.getExecutableComponentDescription(resComp)==null || bOverwrite) {
					// Component does not exist
					// Add to the set to modify before adding to the repository repository
					setComponentsToAdd.add(qrNew.getExecutableComponentDescription(resComp));
					setAddedURIs.add(resComp.toString());
				}
				else {
					// Component does exist
					log.warning("Component "+resComp+" already exists in "+request.getRemoteUser()+"'s repository. Ignoring it.");
				}
			}

			// Adding flows
			Model modUser = qr.getModel();
			modUser.begin();
			for ( Resource resFlow:qrNew.getAvailableFlows() ) {
				if ( qr.getFlowDescription(resFlow)==null ) {
					// Component does not exist
					modUser.add(qrNew.getFlowDescription(resFlow).getModel());
					setAddedURIs.add(resFlow.toString());
					log.info("Added flow "+resFlow+" to "+request.getRemoteUser()+"'s repository");

				}
				else {
					if ( bOverwrite ) {
						log.info("Overwriting flow "+resFlow);
						modUser.remove(qr.getFlowDescription(resFlow).getModel());
						modUser.add(qrNew.getFlowDescription(resFlow).getModel());
						setAddedURIs.add(resFlow.toString());
						log.info("Added flow "+resFlow+" to "+request.getRemoteUser()+"'s repository");
					}
					else
						// Component does exist
						log.warning("Flow "+resFlow+" already exists in "+request.getRemoteUser()+"'s repository. Ignoring it. No overwrite flag provided.");
				}
			}

			// Adding the components after adding the contexts
			Set<String> setFiles = htContextBytes.keySet();
			boolean bWriteOnce = true;
			for ( ExecutableComponentDescription ecd:setComponentsToAdd) {
				if ( htContextBytes.keySet().isEmpty() ) {
					if ( qr.getExecutableComponentDescription(ecd.getExecutableComponent())!=null ) {
						if ( bOverwrite ) {
							modUser.remove(qr.getExecutableComponentDescription(ecd.getExecutableComponent()).getModel());
							modUser.add(ecd.getExecutableComponent().getModel());
							log.info("Added component "+ecd.getExecutableComponent()+" to "+request.getRemoteUser()+"'s repository");
						}
						else
							log.warning("Discarding upload of the existing component "+ecd.getExecutableComponent()+". No overwrite flag provided.");
					}
					else {
						modUser.add(ecd.getModel());
						log.info("Added component "+ecd.getExecutableComponent()+" to "+request.getRemoteUser()+"'s repository");
					}
				}
				else {
					if ( (bEmbed && ecd.getRunnable().equals("java") && ecd.getFormat().equals("java/class") ) ||
							 (ecd.getRunnable().equals("python") && ecd.getFormat().equals("jython"))||
							 (ecd.getRunnable().equals("lisp") && ecd.getFormat().equals("clojure")) ) {
						//
						// Embed all the context per descriptor
						//
						for ( String sFile:setFiles) {
							Literal lit = modUser.createTypedLiteral(htContextBytes.get(sFile),XSDDatatype.XSDbase64Binary);
							ecd.getContext().add(lit);
						}
					}
					else if ( bWriteOnce ) {
                        String javaContextDir = cnf.getPublicResourcesDirectory()+File.separator+"contexts"+File.separator+"java"+File.separator;
                        new File(javaContextDir).mkdirs();

                        if (resManager == null)
                            resManager = (JarToolServlet)PluginFactory.getPluginFactory(cnf).getPlugin("JARTOOL");

					    //
						// Dump the context file to the disc only once
						//
						for ( String sFile:setFiles) {
							// Fail-safe check
							if ( sFile.length()==0 ) continue;

							// Dump the files to disk
							byte[] fileData = htContextBytes.get(sFile);
							if (fileData.length > 0) {     // MDR-160: ignore uploading context files of 0 size
							    // Create the MD5 checksum for the uploaded file (MDR-160)
							    String sMD5 = Crypto.getHexString(Crypto.createMD5Checksum(fileData)).toLowerCase();
							    String sJarFile = resManager.getResourceForMD5(sMD5);

							    if (sJarFile == null) {
							        File savedFile = new File(javaContextDir, sFile);
							        try {
							            FileOutputStream fos = new FileOutputStream(savedFile);
							            fos.write(fileData);
							            fos.close();
							        } catch (Exception e) {
							            log.warning("Problems writing context "+sFile+" to "+savedFile.getAbsolutePath());
							            throw new IOException(e.toString());
							        }

							        try {
                                        resManager.addResource(sFile, sMD5);
                                    }
                                    catch (ResourceConflictException e) {
                                        log.log(Level.WARNING, "JAR naming conflict!", e);
                                    }
							    } else {
							        log.finer("Skipping upload of '" + sFile + "'. Reason: Found as '" + sJarFile + "'");
							        sFile = sJarFile;
                                }
							} else
							    log.finer("Skipping upload of '" + sFile + "'. Reason: File size is 0");

							ecd.getContext().add(modUser.createResource("context://localhost/java/"+sFile));
						}
						// Avoid dumping them multiple times
						bWriteOnce = false;
					}

					if ( qr.getExecutableComponentDescription(ecd.getExecutableComponent())!=null ) {
						if ( bOverwrite ) {
							modUser.remove(qr.getExecutableComponentDescription(ecd.getExecutableComponent()).getModel());
							modUser.add(ecd.getModel());
							log.info("Adding component "+ecd.getExecutableComponent()+" to "+request.getRemoteUser()+"'s repository");
						}
						else
							log.warning("Discarding upload of the existem component "+ecd.getExecutableComponent()+". No overwrite flag provided.");
					}
					else {
						modUser.add(ecd.getModel());
					}
				}

			}

			// Commit changes
			modUser.commit();

			// Regenerate the cache after adding
			qr.refreshCache();
		}

		return setAddedURIs;
	}

}
